### PR TITLE
Safer metadata table handling

### PR DESCRIPTION
**Pull request by vinjana**
* metadataColumnIDs now defines a priority order for the mandatory columns.
* Changed unsafeSubsetByX methods into a safe variant that checks that higher priority columns are unique whenever a subset of rows is selected to have a specific value.

Before these changes the default subsetByX methods easily allowed implementing client code, in which e.g. data from different patients could be yielded, when selecting a specific sample identifier.

### CODE REVIEW

**Comment by vinjana** [comment]
Note that the binary was not updated after the merge of the feature branch with my local development branch.

File: RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy [hunk 1]
```diff
@@ -67,7 +67,7 @@ public class BaseMetadataTable {
     }
 
     BaseMetadataTable(Map<String, Integer> headerMap, Map<String, String> internal2CustomIDMap, List<String> mandatoryColumns, List<Map<String, String>> records) {
-        this.internal2CustomIDMap = internal2CustomIDMap;
+        this.internal2CustomIDMap = internal2CustomIDMap as LinkedHashMap<String,String>;
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy]: Why is this necessary?

File: RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy [hunk 2]
```diff
@@ -156,22 +156,51 @@ public class BaseMetadataTable {
...
+    public BaseMetadataTable subsetByColumn(String columnName, String value) {
+        return unsafeSubsetByColumn(columnName, value).assertUniqueness(columnName)
+    }
+
+
+    /** Given a column names, throw if that column or some higher-priority mandatory column have non-unique values. */
+    public BaseMetadataTable assertUniqueness(String columnName = null) {
+        boolean result = true
+        def mandatoryUniqueColumnNames = mandatoryColumnNames
+        if (!mandatoryColumnNames.contains(columnName) || columnName != INPUT_TABLE_FILE)
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy]: Maybe you can comment, why the exclusion of INPUT_TABLE_FILE is necessary?

**Comment by dankwart-de** [comment]
See file comments

File: RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy (hunk 1 above)
Comment (vinjana) [inline RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy]: Remnant of a first implementation attempt. Will be removed.

**Comment by vinjana** [comment]
Note: The last codefix that you introduced is not yet in the jar.

**Comment by vinjana** [comment]
Note further that for this to run with the COProjects plugin the column order in the metadataTableColumnIDs variable has to be changed to e.g. "datasetCol,mergeCol,markCol,runCol,mateCol,fileCol,readLayoutCol"
Will be committed in the plugin.

File: RoddyCore/src/de/dkfz/roddy/config/ConfigurationFactory.groovy [hunk 3]
```diff
@@ -85,9 +85,9 @@ public class ConfigurationFactory {
         configurationDirectories.parallelStream().each {
             File baseDir ->
                 logger.log(Level.CONFIG, "Searching for configuration files in: " + baseDir.toString());
-                if (!baseDir.canRead()) {
-                    logger.log(Level.SEVERE, "Cannot read from configuration directory ${baseDir.absolutePath}, does the folder exist und do you have access rights to it?")
-                    return;
+                if (!baseDir.canRead() || !baseDir.canExecute()) {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/config/ConfigurationFactory.groovy]: Add TODO to implement a local file system access provider.

File: RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy [hunk 4]
```diff
@@ -193,4 +219,15 @@ public class BaseMetadataTable {
     public List<Map<String, String>> getRecords() {
         return records;
     }
+
+    public BaseMetadataTable unsafeSubsetBy(Map<String, String> columnValueMap) {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/execution/io/BaseMetadataTable.groovy]: Change order of entries.

File: RoddyCore/test/de/dkfz/roddy/execution/io/BaseMetadataTableTest.groovy [hunk 5]
```diff
@@ -93,9 +95,25 @@ public class BaseMetadataTableTest {
...
-        def column = table.subsetByColumn(DATASET_COL, "b")
-        assert column.size() == 2
-        assert table.subsetByColumn(DATASET_COL, "a").size() == 4
+        def subtable = table.subsetByColumn(BaseMetadataTable.INPUT_TABLE_DATASET, "b")
+        assert subtable.size() == 2
+        assert table.subsetByColumn(BaseMetadataTable.INPUT_TABLE_DATASET, "a").size() == 4
+
+        assert table.unsafeSubsetByColumn(RUMPLE_COL, "ambiguous").size() == 8
+        try {
+            table.subsetByColumn(RUMPLE_COL, "ambiguous")
```
Comment (dankwart-de) [inline RoddyCore/test/de/dkfz/roddy/execution/io/BaseMetadataTableTest.groovy]: assert false OR new exception testing

File: RoddyCore/src/de/dkfz/roddy/config/ConfigurationFactory.groovy (hunk 3 above)
Comment (vinjana) [inline RoddyCore/src/de/dkfz/roddy/config/ConfigurationFactory.groovy]: done
